homeassistant/components/daikin.py

```diff
@@ -0,0 +1,294 @@
+
+import voluptuous as vol
+
+from homeassistant.components.climate import (
+    ATTR_OPERATION_MODE, ATTR_SWING_MODE,
+    ATTR_TEMPERATURE, ATTR_CURRENT_TEMPERATURE, ATTR_TARGET_TEMP_STEP,
+    ATTR_FAN_MODE,
+
+    STATE_OFF,
+    STATE_AUTO, STATE_HEAT, STATE_COOL, STATE_DRY, STATE_FAN_ONLY
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.discovery import SERVICE_DAIKIN
+from homeassistant.helpers import discovery
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.const import (
+    CONF_HOSTS, CONF_ICON, CONF_NAME,
+    CONF_MONITORED_CONDITIONS,
+    TEMP_CELSIUS
+)
+
+REQUIREMENTS = ['pydaikin==0.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'daikin'
+HTTP_RESOURCES = ['aircon/get_sensor_info', 'aircon/get_control_info']
+
+ATTR_TARGET_TEMPERATURE = 'target_temperature'
+ATTR_INSIDE_TEMPERATURE = 'inside_temperature'
+ATTR_OUTSIDE_TEMPERATURE = 'outside_temperature'
+
+# default scan interval in seconds
+DEFAULT_SCAN_INTERVAL = 60
+
+COMPONENT_TYPES = ['climate', 'sensor']
+
+SENSOR_TYPES = {
+    ATTR_INSIDE_TEMPERATURE: {
+        CONF_NAME: 'Inside Temperature',
+        CONF_ICON: 'mdi:thermometer'
+    },
+    ATTR_OUTSIDE_TEMPERATURE: {
+        CONF_NAME: 'Outside Temperature',
+        CONF_ICON: 'mdi:thermometer'
+    }
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_HOSTS, default=[]): vol.Schema([cv.string]),
+        vol.Optional(
+            CONF_MONITORED_CONDITIONS,
+            default=list(SENSOR_TYPES.keys())
+        ): vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)])
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+HA_STATE_TO_DAIKIN = {
+    STATE_FAN_ONLY: 'fan',
+    STATE_DRY: 'dry',
+    STATE_COOL: 'cool',
+    STATE_HEAT: 'hot',
+    STATE_AUTO: 'auto',
+    STATE_OFF: 'off',
+}
+
+KNOWN_DEVICES = []
+
+
+def setup(hass, config):
+    """Establish connection with Daikin."""
+    def discovery_dispatch(service, discovery_info):
+        """Dispatcher for Daikin discovery events."""
+        host = discovery_info.get('ip')
+
+        if manual_device_setup(hass, host) is None:
+            return
+
+        for component in COMPONENT_TYPES:
+            load_platform(hass, component, DOMAIN, discovery_info,
+                          config)
+
+    discovery.listen(hass, SERVICE_DAIKIN, discovery_dispatch)
+
+    devices = []
+
+    # Add static devices from the config file.
+    devices.extend(
+        (host, None)
+        for host in config.get(DOMAIN, {}).get(CONF_HOSTS, [])
+    )
+
+    for host in devices:
+        if manual_device_setup(hass, host) is not None:
+            discovery_info = {
+                'ip': host,
+                CONF_MONITORED_CONDITIONS:
+                    config.get(DOMAIN, {}).get(CONF_MONITORED_CONDITIONS)
+            }
+            load_platform(hass, 'sensor', DOMAIN, discovery_info, config)
+
+    return True
+
+
+def manual_device_setup(hass, host, name=None):
```

Comment:
Since you are manipulating shared data structures here, you must watch out for concurrency. This method is called from your `setup_platform` methods. Since they are not written in an async fashion, I think they may be run in executors, i.e., concurrently. In this case, you would be dealing with a race condition here.

I'm not sure about this, though. In any case, doing locking here may be a good idea.

Reply:
The purpose of this functions is to make sure that the climate and sensors all share the same API class, which is DaikinEntity no matter how many Daikin AC devices u have.

I would like to clarify is that the fact that this function is used in all three file:
- platform
- climate 
- sensor files
files in order to make sure that a real Daikin AC device is not added multiple times.

I have 2 different Daikin AC with 2 different controller versions and I've tested mixed (if not all) cases:
- automatic discovery which finds both devices and adds all sensors
- one manual climate + one automatic discovery of the other device
- manual climates + manual sensors
- etc
And I've never encountered any race conditions or any errors in the logs. Actually what happend is the fact that manual configuration ads climate and sensors instantly while discovery takes a while makes sure that it doesn't add any previously added device (see KNOWN_HOSTS variable)

If u still feel that this is wrong I'm opened to any solution on this since I've said that I'm a beginner with HA and also python development (not to mention multi-threading).

Reply:
@tinloaf can u help me with some updates on this PR ?